### PR TITLE
Revision aggregate

### PR DESCRIPTION
**Pull request by sabin95**

### CODE REVIEW

File: BE.BL/Cars/Create/CarCreateCommand.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+namespace BE.BL.Cars.Create
+{
+    public class CarCreateCommand
```
Comment (PiMihai) [inline BE.BL/Cars/Create/CarCreateCommand.cs]:
```suggestion
    public class CreateCarCommand
```

File: BE.BL/Cars/Edit/CarEditByIdCommand.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+namespace BE.BL.Cars.Edit
+{
+    public class CarEditByIdCommand
```
Comment (PiMihai) [inline BE.BL/Cars/Edit/CarEditByIdCommand.cs]:
```suggestion
    public class EditCarCommand
```

File: BE.BL/Revisions/Create/RevisionCreateCommand.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+namespace BE.BL.Revisions.Create
+{
+    public class RevisionCreateCommand
```
Comment (PiMihai) [inline BE.BL/Revisions/Create/RevisionCreateCommand.cs]:
```suggestion
    public class CreateRevisionCommand
```

File: BE.BL/Revisions/Edit/RevisionEditByIdCommand.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+namespace BE.BL.Revisions.Edit
+{
+    public class RevisionEditByIdCommand
```
Comment (PiMihai) [inline BE.BL/Revisions/Edit/RevisionEditByIdCommand.cs]:
```suggestion
    public class EditRevisionCommand
```

File: BE.Queries/Cars/GetAllCarsByClientId/GetAllCarsByClientIdResult.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+namespace BE.Queries.Cars
+{
+    public class GetAllCarsByClientIdResult
```
Comment (PiMihai) [inline BE.Queries/Cars/GetAllCarsByClientId/GetAllCarsByClientIdResult.cs]:
```suggestion
    public class GetAllCarsForClientResult
```

File: BE.Queries/Cars/GetCarById/GetCarByIdResult.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+namespace BE.Queries.Cars
+{
+    public class GetCarByIdResult
```
Comment (PiMihai) [inline BE.Queries/Cars/GetCarById/GetCarByIdResult.cs]:
```suggestion
    public class GetCarResult
```

File: BE.Queries/Revisions/GetAllRevisionsByClientId/GetAllRevisionsByClientIdResult.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+namespace BE.Queries.Revisions.GetAllRevisionsByClientId
+{
+    public class GetAllRevisionsByClientIdResult
```
Comment (PiMihai) [inline BE.Queries/Revisions/GetAllRevisionsByClientId/GetAllRevisionsByClientIdResult.cs]:
```suggestion
    public class GetRevisionsForClientResult
```

File: BE.Queries/Revisions/GetRevisionById/GetRevisionByIdResult.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+namespace BE.Queries.Revisions.GetRevisionById
+{
+    public class GetRevisionByIdResult
```
Comment (PiMihai) [inline BE.Queries/Revisions/GetRevisionById/GetRevisionByIdResult.cs]:
```suggestion
    public class GetRevisionResult
```

File: BE/Controllers/ClientsController.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+
+namespace BE.Controllers
+{
+    [Route("api/[controller]")]
+    public class ClientsController
+    {
+
+        #region Ctor
```
Comment (PiMihai) [inline BE/Controllers/ClientsController.cs]: Never **ever** use #region ever even if your life depends on it.

File: BE/Controllers/ClientsController.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        public List<GetAllRevisionsByClientIdResult> GetAllRevisionsByClientId(long clientId)
+        {
+
+            return _context.Set<GetAllRevisionsByClientIdResult>().FromSqlRaw("EXEC [dbo].[usp_GetAllRevisionsByClientId] {0}", clientId).ToList();
+        }
+
+        #endregion
+
+        #region Post
+        [HttpPost("{clientId}/Cars")]
```
Comment (PiMihai) [inline BE/Controllers/ClientsController.cs]: REST paths should always be **camelCase**

```suggestion
        [HttpPost("{clientId}/cars")]
```

File: BE/Controllers/ClientsController.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        public void AddCar(long clientId,[FromBody] CarCreateCommand carCreateCommand)
+        {
+            _context.Database.ExecuteSqlRaw("EXEC [dbo].[usp_InsertCar] {0},{1},{2},{3},{4}", 
+                                                clientId,carCreateCommand.BrandName,
+                                                carCreateCommand.ModelName,carCreateCommand.PlateNumber,
+                                                carCreateCommand.RegistrationId);
+        }
+
+        [HttpPost("{clientId}/Revisions")]
```
Comment (PiMihai) [inline BE/Controllers/ClientsController.cs]:
```suggestion
        [HttpPost("{clientId}/revisions")]
```

File: BE/Program.cs [hunk 12]
```diff
@@ -15,8 +15,9 @@ public static void Main(string[] args)
                 .ConfigureWebHostDefaults(webBuilder =>
                 {
                     webBuilder
-                        .UseUrls("http://0.0.0.0:5000", "https://0.0.0.0:5001")
+                        //.UseUrls("http://0.0.0.0:5000", "https://0.0.0.0:5001")
```
Comment (PiMihai) [inline BE/Program.cs]: Delete commented code.

File: BE/Startup.cs [hunk 13]
```diff
@@ -1,67 +1,91 @@
...
+using BE.BL.Revisions.Create;
+using BE.BL.Revisions.Edit;
+using BE.Queries.Cars;
+using BE.BL.Cars.Create;
+using BE.BL.Cars.Edit;
+using BE.Queries.Revisions.GetAllRevisionsByClientId;
 
 namespace BE
 {
     public class MyContext : DbContext
```
Comment (PiMihai) [inline BE/Startup.cs]: Rename to something more interesting, maybe SqlDbContext or GetYourCarFixedDbContext

File: BE/Startup.cs [hunk 14]
```diff
@@ -1,67 +1,91 @@
...
         protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
             modelBuilder.Entity<MyClass>().HasNoKey();
+            modelBuilder.Entity<GetAllRevisionsResult>().HasNoKey();
+            modelBuilder.Entity<GetAllRevisionsByClientIdResult>().HasNoKey();
+            modelBuilder.Entity<GetRevisionByIdResult>().HasNoKey();
+            modelBuilder.Entity<RevisionCreateCommand>().HasNoKey();
+            modelBuilder.Entity<RevisionEditByIdCommand>().HasNoKey();
+            modelBuilder.Entity<CarCreateCommand>().HasNoKey();
+            modelBuilder.Entity<CarEditByIdCommand>().HasNoKey();
```
Comment (PiMihai) [inline BE/Startup.cs]: You only need to use this for classes that read from SQL query outputs (like results). Commands never reach EntityFramework (only their properties do but that's fine).
```suggestion
```

File: BE/Startup.cs [hunk 15]
```diff
@@ -1,67 +1,91 @@
...
+            modelBuilder.Entity<GetAllCarsByClientIdResult>().HasNoKey();
+            modelBuilder.Entity<GetCarByIdResult>().HasNoKey();
         }
     }
 
     public class MyClass
```
Comment (PiMihai) [inline BE/Startup.cs]: Is this still needed?

File: DB/StoredProcedures/DeleteCarById.sql [hunk 16]
```diff
@@ -0,0 +1,6 @@
+CREATE OR ALTER PROCEDURE usp_DeleteCarById @Id BIGINT, @ClientId BIGINT AS
```
Comment (PiMihai) [inline DB/StoredProcedures/DeleteCarById.sql]: Formatting tip : 

```suggestion
CREATE OR ALTER PROCEDURE usp_DeleteCarById
    @Id BIGINT
  , @ClientId BIGINT
AS
```

File: DB/StoredProcedures/DeleteRevisionById.sql [hunk 17]
```diff
@@ -0,0 +1,6 @@
+CREATE OR ALTER PROCEDURE usp_DeleteRevisionById @Id BIGINT AS 
+BEGIN
+    Delete
+    FROM dbo.Revisions
+    Where Id=@Id
```
Comment (PiMihai) [inline DB/StoredProcedures/DeleteRevisionById.sql]: Also bind to Client and Car to prevent misuse.

File: DB/StoredProcedures/GetAllCarsByClientId.sql [hunk 18]
```diff
@@ -0,0 +1,6 @@
+CREATE OR ALTER PROCEDURE usp_GetAllCarsByClientId @ClientId BIGINT AS 
+BEGIN
+    SELECT Id,ClientId,BrandName,ModelName,PlateNumber,RegistrationId
```
Comment (PiMihai) [inline DB/StoredProcedures/GetAllCarsByClientId.sql]: Formatting tip :
```suggestion
    SELECT
        Id
      , ClientId
      , BrandName
      , ModelName
      , PlateNumber
      , RegistrationId
```

File: DB/StoredProcedures/GetAllStatus.sql [hunk 19]
```diff
@@ -0,0 +1,5 @@
+CREATE OR ALTER PROCEDURE usp_GetAllStatus AS
```
Comment (PiMihai) [inline DB/StoredProcedures/GetAllStatus.sql]: This is not used anywhere in the app.

File: DB/StoredProcedures/GetRevisionById.sql [hunk 20]
```diff
@@ -0,0 +1,8 @@
+CREATE OR ALTER PROCEDURE usp_GetRevisionById @Id BIGINT AS 
+BEGIN
+    SELECT r.Id,r.ClientId,r.Title,r.ProblemDetails,c.BrandName,c.ModelName,c.PlateNumber,c.RegistrationId
+    FROM dbo.Revisions r
+    JOIN dbo.Cars  c
+    ON c.Id = r.CarId
+    Where r.Id=@Id
```
Comment (PiMihai) [inline DB/StoredProcedures/GetRevisionById.sql]: Also bind to Client and Car to prevent misuse.

File: FE/src/app/car-insert/CarCreateCommand.ts [hunk 21]
```diff
@@ -0,0 +1,6 @@
+export interface CarCreateCommand{
```
Comment (PiMihai) [inline FE/src/app/car-insert/CarCreateCommand.ts]:
```suggestion
export interface CreateCarCommand {
```

File: FE/src/app/car-insert/car-insert.component.ts [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  async onSubmit() {
+    let car = <CarCreateCommand> {
+      brandName:this.brandName,
+      modelName:this.modelName,
+      plateNumber:this.plateNumber,
+      registrationId:this.registrationId
+    }
+    await this.http.post("http://localhost:5000/api/clients/1/Cars", car).toPromise();
```
Comment (PiMihai) [inline FE/src/app/car-insert/car-insert.component.ts]: Create a separate service with a method that does this and inject that instead of httpClient.

File: FE/src/app/revision/RevisionCreateCommand.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+export interface RevisionCreateCommand {
```
Comment (PiMihai) [inline FE/src/app/revision/RevisionCreateCommand.ts]:
```suggestion
export interface CreateRevisionCommand {
```

File: FE/src/app/revision/revision.component.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+  problemDetails:string
+  title:string
+  
+  
+  constructor(
+    public http:HttpClient
+  ) { }
+  
+  async ngOnInit() {
+    this.cars = await this.http.get<GetCarResult>("http://localhost:5000/api/clients/1/cars").toPromise()
```
Comment (PiMihai) [inline FE/src/app/revision/revision.component.ts]: Move this to service.

File: FE/src/app/revision/revision.component.ts [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  async onSubmit() {
+    console.dir(this.selectedCarId);
+    let revision = <RevisionCreateCommand> {
+      carId:+this.selectedCarId,
+      title:this.title,
+      problemDetails:this.problemDetails
+    }
+    await this.http.post("http://localhost:5000/api/clients/1/Revisions", revision).toPromise();
```
Comment (PiMihai) [inline FE/src/app/revision/revision.component.ts]: Move this to service.

File: FE/src/app/revisions-list/revisions-list.component.ts [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+  styleUrls: ['./revisions-list.component.scss']
+})
+export class RevisionsListComponent implements OnInit {
+  revisions: GetRevisionResult
+  constructor(
+    public http: HttpClient
+  ) { }
+
+  async ngOnInit() {
+    this.revisions = await this.http.get<GetRevisionResult>("http://localhost:5000/api/clients/1/revisions").toPromise()
```
Comment (PiMihai) [inline FE/src/app/revisions-list/revisions-list.component.ts]: Move this to service.

File: FE/src/index.html [hunk 27]
```diff
@@ -6,8 +6,10 @@
   <base href="/">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
+  <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500&display=swap" rel="stylesheet">
+  <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
 </head>
-<body>
+<body class="mat-typography">
```
Comment (PiMihai) [inline FE/src/index.html]: is `class="mat-typography"` still needed?

**Review by PiMihai — CHANGES_REQUESTED** [review]
Looks good overall, just a few naming changes & small refactorings.

Things to think about : 
* where is the RevisionAggregate? (the Pull Request is named that way)
* in what order should you run the SQL scripts on a blank database?
* could batching be useful? (GetXBy**Ids**)

File: BE/Startup.cs (hunk 15 above)
Comment (sabin95) [reply BE/Startup.cs]: no

File: DB/StoredProcedures/GetAllStatus.sql (hunk 19 above)
Comment (sabin95) [reply DB/StoredProcedures/GetAllStatus.sql]: yet

File: DB/StoredProcedures/DeleteRevisionById.sql (hunk 17 above)
Comment (sabin95) [reply DB/StoredProcedures/DeleteRevisionById.sql]: done - but will implement soft delete here

File: FE/src/index.html (hunk 27 above)
Comment (sabin95) [reply FE/src/index.html]: nope

File: BE/BE.csproj [hunk 28]
```diff
@@ -5,7 +5,6 @@
     <UserSecretsId>9c0f0207-84a3-4911-9ba5-c31596c3c354</UserSecretsId>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.2.8" />
```
Comment (PiMihai) [inline BE/BE.csproj]: Why did you remove this reference?

**Review by PiMihai — APPROVED** [review]
